edgelet/hsm-sys/azure-iot-hsm-c/src/edge_hsm_client_x509.c

```diff
@@ -1,72 +1,501 @@
+        {
+            g_is_x509_initialized = true;
+        }
+    }
+    else
+    {
+        result = 0;
+    }
+
+    return result;
 }
 
 void hsm_client_x509_deinit()
 {
+    if (!g_is_x509_initialized)
+    {
+        LOG_ERROR("hsm_client_x509_init not called");
+    }
+    else
+    {
+        if (g_ref_cnt == 0)
+        {
+            g_is_x509_initialized = false;
+            hsm_client_crypto_deinit();
+        }
+    }
 }
 
-void iothub_hsm_free_buffer(void * buffer)
+void edge_x509_hsm_free_buffer(void * buffer)
 {
     if (buffer != NULL)
     {
         free(buffer);
     }
 }
 
-HSM_CLIENT_HANDLE iothub_x509_hsm_create()
+HSM_CLIENT_HANDLE edge_x598_hsm_create()
 {
-    LOG_ERROR("API unsupported");
-    return (HSM_CLIENT_HANDLE)NULL;
+    HSM_CLIENT_HANDLE result;
+
+    if (!g_is_x509_initialized)
+    {
+        LOG_ERROR("hsm_client_x509_init not called");
+        result = NULL;
+    }
+    else
+    {
+        const HSM_CLIENT_CRYPTO_INTERFACE* interface = hsm_client_crypto_interface();
+        if (interface == NULL)
+        {
+            LOG_ERROR("hsm_client_crypto_interface returned NULL");
+            result = NULL;
+        }
+        else
+        {
+            result = interface->hsm_client_crypto_create();
+            if (result != NULL)
+            {
+                g_ref_cnt++;
+            }
+        }
+    }
+
+    return result;
 }
 
-void iothub_x509_hsm_destroy(HSM_CLIENT_HANDLE handle)
+void edge_x509_hsm_destroy(HSM_CLIENT_HANDLE hsm_handle)
 {
-    (void)handle;
-    LOG_ERROR("API unsupported");
+    if (!g_is_x509_initialized)
+    {
+        LOG_ERROR("hsm_client_x509_init not called");
+    }
+    else
+    {
+        if (hsm_handle == NULL)
+        {
+            LOG_ERROR("Null hsm handle parameter");
+        }
+        else if (g_ref_cnt == 0)
+        {
+            LOG_ERROR("Mismatch in overall handle create and destroy calls");
+        }
+        else
+        {
+            const HSM_CLIENT_CRYPTO_INTERFACE* interface = hsm_client_crypto_interface();
+            if (interface == NULL)
+            {
+                LOG_ERROR("hsm_client_crypto_interface returned NULL");
+            }
+            else
+            {
+                interface->hsm_client_crypto_destroy(hsm_handle);
+            }
+            g_ref_cnt--;
+        }
+    }
+}
+
+static CERT_PROPS_HANDLE create_edge_device_properties
+(
+    const char* common_name,
+    uint64_t validity_seconds,
+    const char* issuer_alias
+)
+{
+    CERT_PROPS_HANDLE certificate_props = cert_properties_create();
+    const char* alias = EDGE_DEVICE_ALIAS;
+
+    if (certificate_props == NULL)
+    {
+        LOG_ERROR("Could not create certificate props for %s", alias);
+    }
+    else if (set_common_name(certificate_props, common_name) != 0)
+    {
+        LOG_ERROR("Could not set common name for %s", alias);
+        cert_properties_destroy(certificate_props);
+        certificate_props = NULL;
```

Comment:
I know this is putting salt in the wound, but this function could be so much cleaner with `goto error;` :)